src/renderers/common/StorageInstancedBufferAttribute.js

```diff
@@ -10,6 +10,99 @@ class StorageInstancedBufferAttribute extends InstancedBufferAttribute {
 
 		this.isStorageInstancedBufferAttribute = true;
 
+		this.typeClass = typeClass;
+
+		this.isStructured = false;
+
+		this.isStructArray = false;
+
+		this.structs = [];
+
+	}
+
+	setStructs( structs ) {
+
+		this.structs = structs.map( ( struct ) => {
+
+			//Without specifying the type, itemSize and typeClass are used to set the default type
+			const type = struct.type || this.getDefaultType( this.itemSize );
+
+			return {
+
+				name: struct.name,
+				type: type,
+				isAtomic: struct.isAtomic !== undefined ? struct.isAtomic : false
+
+			};
+
+		} );
+
+		this.isStructured = this.structs.length > 0;
+
+		if ( this.count === this.structs.length ) {
+
+			this.isStructArray = false;	//is single struct
+
+		} else if ( this.count % this.structs.length !== 0 ) {
+
+			throw new Error( 'Structuring does not match the storage array length' );
+
+		} else {
+
+			this.isStructArray = true;	//is structArray
+
+		}
+
+	}
+
+	getDefaultType( itemSize ) {
+
+		let typePrefix = '';
+
+		if ( this.typeClass === Uint32Array ) {
+
+			typePrefix = 'u';
+
+		} else if ( this.typeClass === Int32Array ) {
+
+			typePrefix = 'i';
+
+		} else if ( this.typeClass === Float32Array ) {
+
+			typePrefix = 'f';
+
+		}
+
+		switch ( itemSize ) {
+
+			case 4: return `vec4<${ typePrefix }32>`;
+			case 3: return `vec3<${ typePrefix }32>`;
+			case 2: return `vec2<${ typePrefix }32>`;
+			case 1: return `${ typePrefix }32`;
+			default:
+				throw new Error( `Unsupported itemSize: ${ itemSize }. Expected 1, 2, 3, or 4.` );
+
+		}
+
+	}
+
+	getStructs() {
+
+		return this.structs;
+
+	}
+
+	getWGSLStructString( structName = this.name ) {
```

Comment:
Complex snippets are always generated by `NodeBuilder`, i.e. `WGSLNodeBuilder` in this case.